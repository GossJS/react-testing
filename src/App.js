import React, { Component } from 'react';
import AppBar from 'material-ui/AppBar';
import RaisedButton from 'material-ui/RaisedButton';
import Paper from 'material-ui/Paper';
import TextField from 'material-ui/TextField';
import CircularProgress from 'material-ui/CircularProgress';
import PropTypes from 'prop-types';

import style from './App.style.js';

export default class App extends Component {
    constructor(props) {
        super(props);

        this.state = {
            text: '',
            imageUrl: '',
            isImageLoading: false,
        };
        this.handleOnChangeText = this.handleOnChangeText.bind(this);
        this.handleOnClickBtn = this.handleOnClickBtn.bind(this);
    }

    handleOnChangeText(event) {
        this.setState({ text: event.target.value });
    }

    handleOnClickBtn() {
        this.setState({ isImageLoading: true });
        this.fetchImage();
    }

    fetchImage() {
        return fetch(`https://picsum.photos/${this.props.width}/${this.props.height}/?random`)
            .then(resolve => resolve.blob())
            .then(resolve => this.setState({ imageUrl: window.URL.createObjectURL(resolve), isImageLoading: false }));
    }

    render() {
        return (
            <main>
                <AppBar
                    title="Put text on random pic"
                    iconElementLeft={ <div></div> }
                />

                <Paper style={ style.paper } zDepth={ 4 }>
                    {
                        this.state.isImageLoading ?
                            <CircularProgress style={ style.spinner } /> :
                            null
                    }
                    <div style={ style.picture }>
                        <img src={ this.state.imageUrl } alt=""/>
                        {
                            this.state.imageUrl ?
                                <span style={ style.text }>{ this.state.text }</span> :
                                null
                        }
                    </div>

                    <div style={ style.controls }>
                        <TextField
                            hintText="your text"
                            style={ { marginRight: '15px' } }
                            onChange={ this.handleOnChangeText }
                        />
                        <RaisedButton
                            label="load new pic"
                            onClick={ this.handleOnClickBtn }
                        />
                    </div>
                </Paper>
            </main>
        );
    }
}

App.defaultProps = {
    width: 400,
    height: 300,
};

App.propTypes = {
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
};
