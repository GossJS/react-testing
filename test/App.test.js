import React from 'react';
import { shallow } from 'enzyme';

import App from '../src/App';
import AppBar from 'material-ui/AppBar';
import RaisedButton from 'material-ui/RaisedButton';
import Paper from 'material-ui/Paper';
import TextField from 'material-ui/TextField';
import CircularProgress from 'material-ui/CircularProgress';
import "whatwg-fetch";

import jsdom from 'jsdom';

const {JSDOM} = jsdom;
const {document} = (new JSDOM('<!doctype html><html><body></body></html>')).window;
global.document = document;
global.window = document.defaultView;

describe('<App />', () => {
    const props = {
        width: 1200,
        height: 400,
    }
    let wrapper = shallow(<App { ...props } />);

    it('exists', () => {
        expect(wrapper.exists()).toEqual(true);
    });

    it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('receives props', () => {
        expect(wrapper.instance().props.width).toEqual(1200);
    });

    it('contains components', () => {
        expect(wrapper.find(AppBar).length).toEqual(1);
        expect(wrapper.find(RaisedButton).length).toEqual(1);
        expect(wrapper.find(Paper).length).toEqual(1);
        expect(wrapper.find(TextField).length).toEqual(1);
    });

    it('receives text in input but shows no text while there is no image', () => {
        const input = wrapper.find('input');
        input.value = 'some words';
        expect(wrapper.find('span').length).toEqual(0);
    });

    it('sets text in state but shows no text while there is no image', () => {
        wrapper.setState({ text: 'some words'});
        expect(wrapper.find('span').length).toEqual(0);
    });

    it('receives text in input and sets new state', () => {
        const input = wrapper.find('input');
        input.value = 'some words';
        expect(wrapper.state().text).toEqual('some words');
    });

    // it('clicks button', async () => {
    //     const button = wrapper.find(RaisedButton);
    //     button.simulate('click');
    //     expect(wrapper.state().isImageLoading).toEqual(true);
    //     await expect(wrapper.instance().fetchImage()).resolves.toEqual();
    //     // expect(wrapper.state().isImageLoading).toEqual(false);
    //     // expect(wrapper.state().imageUrl).toEqual('');
    // });
});
