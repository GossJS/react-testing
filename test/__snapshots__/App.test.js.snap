// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<main>
  <AppBar
    iconElementLeft={<div />}
    showMenuIconButton={true}
    title="Put text on random pic"
    zDepth={1}
  />
  <Paper
    circle={false}
    rounded={true}
    style={
      Object {
        "margin": "30px auto 0",
        "padding": "25px",
        "width": "1250px",
      }
    }
    transitionEnabled={true}
    zDepth={4}
  >
    <div
      style={
        Object {
          "color": "floralwhite",
          "fontSize": "70px",
          "fontWeight": "bold",
          "height": "400px",
          "position": "relative",
          "textAlign": "center",
          "width": "1200px",
        }
      }
    >
      <img
        alt=""
        src=""
      />
    </div>
    <div
      style={
        Object {
          "padding": "20px 0 0 0",
          "textAlign": "center",
        }
      }
    >
      <TextField
        disabled={false}
        floatingLabelFixed={false}
        fullWidth={false}
        hintText="your text"
        multiLine={false}
        onChange={[Function]}
        rows={1}
        style={
          Object {
            "marginRight": "15px",
          }
        }
        type="text"
        underlineShow={true}
      />
      <RaisedButton
        disabled={false}
        fullWidth={false}
        label="load new pic"
        labelPosition="after"
        onClick={[Function]}
        primary={false}
        secondary={false}
      />
    </div>
  </Paper>
</main>
`;
